domain/secret/state/state.go

```diff
@@ -2534,14 +2562,195 @@ AND  (
 	return revisionResult, nil
 }
 
-// InitialWatchStatementForObsoleteRevision returns the initial watch statement and the table name for watching obsolete revisions.
+type dbrevisionUUIDs []revisionUUID
+
+// InitialWatchStatementForConsumedSecretsChange returns the initial watch statement and the table name for watching consumed secrets.
+func (st State) InitialWatchStatementForConsumedSecretsChange(unitName string) (string, eventsource.NamespaceQuery) {
+	queryFunc := func(ctx context.Context, runner coredatabase.TxnRunner) ([]string, error) {
+		q := `
+SELECT
+    DISTINCT sr.uuid AS &revisionUUID.uuid
+FROM secret_unit_consumer suc
+LEFT JOIN unit u ON u.uuid = suc.unit_uuid
+LEFT JOIN secret_revision sr ON sr.secret_id = suc.secret_id
+WHERE u.unit_id = $unit.unit_id
+GROUP BY sr.secret_id
+HAVING suc.current_revision < MAX(sr.revision)`
+
+		queryParams := []any{
+			unit{UnitName: unitName},
+		}
+
+		stmt, err := st.Prepare(q, append([]any{revisionUUID{}}, queryParams...)...)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		var revUUIDs dbrevisionUUIDs
+		err = runner.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+			err := tx.Query(ctx, stmt, queryParams...).GetAll(&revUUIDs)
+			if errors.Is(err, sqlair.ErrNoRows) {
+				// No consumed secrets found.
+				return nil
+			}
+			return errors.Trace(err)
+		})
+		if err != nil {
+			return nil, errors.Trace(domain.CoerceError(err))
+		}
+
+		result := make([]string, len(revUUIDs))
+		for i, rev := range revUUIDs {
+			result[i] = rev.UUID
+		}
+		return result, nil
+	}
+	return "secret_revision", queryFunc
+}
+
+// GetConsumedSecretURIsWithChanges returns the URIs of the secrets consumed by the specified unit that has new revisions.
+func (st State) GetConsumedSecretURIsWithChanges(ctx context.Context, unitName string, revisionIDs ...string) ([]string, error) {
+	db, err := st.DB()
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	q := `
+SELECT
+    DISTINCT suc.secret_id AS &secretUnitConsumer.secret_id
+FROM secret_unit_consumer suc
+LEFT JOIN unit u ON u.uuid = suc.unit_uuid
+LEFT JOIN secret_revision sr ON sr.secret_id = suc.secret_id
```

Comment:
inner joins - a secret unit consumer row has not null FK to unit and revision tables